src/NuGet.Services.AzureSearch/Owners2AzureSearch/Owners2AzureSearchCommand.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace NuGet.Services.AzureSearch.Owners2AzureSearch
+{
+    public class Owners2AzureSearchCommand
+    {
+        private readonly IDatabaseOwnerFetcher _databaseOwnerFetcher;
+        private readonly IOwnerDataClient _ownerDataClient;
+        private readonly IOwnerSetComparer _ownerSetComparer;
+        private readonly IOwnerIndexActionBuilder _indexActionBuilder;
+        private readonly Func<IBatchPusher> _batchPusherFactory;
+        private readonly IOptionsSnapshot<AzureSearchJobConfiguration> _options;
+        private readonly ILogger<Owners2AzureSearchCommand> _logger;
+
+        public Owners2AzureSearchCommand(
+            IDatabaseOwnerFetcher databaseOwnerFetcher,
+            IOwnerDataClient ownerDataClient,
+            IOwnerSetComparer ownerSetComparer,
+            IOwnerIndexActionBuilder indexActionBuilder,
+            Func<IBatchPusher> batchPusherFactory,
+            IOptionsSnapshot<AzureSearchJobConfiguration> options,
+            ILogger<Owners2AzureSearchCommand> logger)
+        {
+            _databaseOwnerFetcher = databaseOwnerFetcher ?? throw new ArgumentNullException(nameof(databaseOwnerFetcher));
+            _ownerDataClient = ownerDataClient ?? throw new ArgumentNullException(nameof(ownerDataClient));
+            _ownerSetComparer = ownerSetComparer ?? throw new ArgumentNullException(nameof(ownerSetComparer));
+            _indexActionBuilder = indexActionBuilder ?? throw new ArgumentNullException(nameof(indexActionBuilder));
+            _batchPusherFactory = batchPusherFactory ?? throw new ArgumentNullException(nameof(batchPusherFactory));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            if (_options.Value.MaxConcurrentBatches <= 0)
+            {
+                throw new ArgumentException(
+                    $"The {nameof(AzureSearchJobConfiguration.MaxConcurrentBatches)} must be greater than zero.",
+                    nameof(options));
+            }
+        }
+
+        public async Task ExecuteAsync()
```

Comment:
>ExecuteAsync [](start = 26, length = 12)

Should we track the duration of `ExecuteAsync` with a metric? This would help us drive decisions like how frequently we should update owners.

Reply:
The jobs infra traces this duration per loop.

Tracked here https://github.com/NuGet/NuGetGallery/issues/6664.